aws-java-sdk-sts/src/main/java/com/amazonaws/auth/STSSessionCredentialsProvider.java

```diff
@@ -119,36 +160,119 @@ public STSSessionCredentialsProvider(AWSCredentialsProvider longLivedCredentials
      */
     public void setSTSClientEndpoint(String endpoint) {
         securityTokenService.setEndpoint(endpoint);
-        sessionCredentials = null;
+        sessionCredentials.set(null);
     }
 
     @Override
     public AWSSessionCredentials getCredentials() {
         if (needsNewSession()) startSession();
 
-        return sessionCredentials;
+        return sessionCredentials.get().sessionCredentials;
     }
 
     @Override
     public void refresh() {
-        startSession();
+        asyncNewSession();
```

Comment:
Sorry I meant synchronous not synchronized. My concern is that by doing the refresh asynchronously we are violating the contract of the interface. Refresh is more forceful and as a developer I'd expect the refresh to be completed when this method returns and not just queued up to be refreshed at some later time.